### PR TITLE
Algorithm implementation, partitioning, replication, database models and corrections

### PR DESCRIPTION
**Pull request by LuisaMG01**
- Developed the GetDataNodesForUpload method to facilitate the selection of data nodes for file uploads.
- Implemented a weighted random selection strategy to distribute chunks among active data nodes, ensuring efficient utilization of storage.
- Created the MetaData model to encapsulate file metadata, including the file name, size, owner, and associated blocks.
- The GetDataNodesForUpload method saves metadata in the metaData collection after a successful data node selection.
- The UploadFile method manages the process of uploading a specified file in chunks to available data nodes. It handles the file's chunking, retrieves data nodes for storage, and sends each chunk to the designated nodes.
- The SendFile method is responsible for receiving a file chunk from the client and storing it on the data node. It manages the directory structure for user-specific file storage and updates the data node's available capacity.

### CODE REVIEW

File: main_client.py [hunk 1]
```diff
@@ -76,4 +76,12 @@ def main(username, password):
     #client.DownloadFile('test1.txt')
 
 
-    """
+"""
+
```
Comment (santiagoneusa) [inline main_client.py:80]: why is this function commented?
if needed for later development, then delete the unnecessary blank spaces

File: main_client.py [hunk 2]
```diff
@@ -76,4 +76,12 @@ def main(username, password):
...
+if __name__ == '__main__':
+    client = Client(CLIENT_IP, CLIENT_PORT, SERVER_IP, SERVER_PORT)
+    print('Calling Register method...')
+    client.Register(username='JuanM04121', password='Test123')
+    input('Press Enter to continue...')
+    client.UploadFile('test.txt')
+    #client.DownloadFile('test1.txt')
```
Comment (santiagoneusa) [inline main_client.py:87]: delete comment

File: src/data_node/data_node.py [hunk 3]
```diff
@@ -27,12 +28,33 @@ def __init__(
...
+            f.write(request.chunk_data)
+
+        chunk_size_MB = len(request.chunk_data) / MB_IN_BYTES
+        print(f"Received chunk {request.chunk_number} of size {chunk_size_MB} MB")
+        print(f"Capacity before: {self.capacity_MB} MB")
+        self.capacity_MB -= chunk_size_MB
+        print(f"Capacity after: {self.capacity_MB} MB")
+        
+        file_size = os.path.getsize(chunk_file_path)
         return data_node_pb2.Reply(length=file_size)
```
Comment (santiagoneusa) [inline src/data_node/data_node.py:55]: In other functions, the return statement has one space between itself and the line written before.

File: src/models/namenode.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from pydantic import BaseModel
+from typing import List
+
+class Block(BaseModel):
+    Master: str
+    Slaves: List[str]
+
+class MetaData(BaseModel):
```
Comment (santiagoneusa) [inline src/models/namenode.py:8]: between the classes there has to be two spaces. run flak8 or black to format the code.

File: src/name_node/name_node.py [hunk 5]
```diff
@@ -1,11 +1,15 @@
 from src.rpc.name_node import name_node_pb2_grpc, name_node_pb2
+
 import grpc
 import json
 from concurrent import futures
+import random
+
 from src.models.datanode import DataNode
```
Comment (santiagoneusa) [inline src/name_node/name_node.py:8]: dont put spaces between the imports.

**Review by santiagoneusa — APPROVED** [review]

File: src/client/client.py
Comment (JuanM0412) [inline src/client/client.py:1]: A lot of white spaces

File: src/name_node/name_node.py
Comment (JuanM0412) [inline src/name_node/name_node.py:1]: A lot of white spaces

**Review by JuanM0412 — APPROVED** [review]

File: main_client.py (hunk 2 above)
Comment (mahoyos) [reply main_client.py:87]: Borra ese registro quemado. Ya no debería ir ahí.

**Review by mahoyos — CHANGES_REQUESTED** [review]
Solo falta corregir las sugerencias pequeñas de formato en los comentarios y eliminar ese registro quemado. Hacer un commit con esas correcciones para proceder a aprobar

**Review by srestrep74 — APPROVED** [review]

**Pull request merged**